File: WebApplication2/Controllers/PersonsController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using CSharpWebApiSample.Contracts;
using CSharpWebApiSample.Domain;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Logging;

namespace WebApplication2.Controllers
{
    [ApiController]
    [Route("api")]
    public class PersonsController : ControllerBase
    {

        private readonly ILogger<PersonsController> logger;
        private readonly PersonsRepository personsRepository;

        public PersonsController(ILogger<PersonsController> logger, PersonsRepository personsRepository)
        {
            this.logger = logger;
            this.personsRepository = personsRepository;
        }

        [HttpGet]
        [Route("persons/{id}")]
        public async Task<IActionResult> Get(int id)
        {
            var result = await personsRepository.GetPersonByIdAsync(id);

            if(result is null)
            {
                return NotFound(id);
            }

            return Ok(result);
        }

        [HttpPost]
        [Route("persons")]
        public async Task<IActionResult> InsertPerson([FromBody] PersonDto person)
        {
            var result = await personsRepository.InsertPerson(person);

            return Ok(result);
        }

        [HttpPut]
        [Route("persons")]
        public async Task<IActionResult> UpdatePerson([FromBody] PersonDto person)
        {
            await personsRepository.UpdatePerson(person);

            return Ok(person);
        }

        [HttpDelete]
        [Route("persons/{id}")]
        public async Task<IActionResult> DeletePerson(int id)
        {
            await personsRepository.DeletePerson(id);

            return Ok(id);
        }
    }
}

File: FSharp/Controllers/PersonsController.fs
namespace Controllers

open Microsoft.AspNetCore.Mvc
open Services

[<ApiController>]
[<Route("api")>]
type PersonsController(personsRepository: PersonsRepository) as this =
    inherit ControllerBase()

    [<Route("persons/{id}")>]
    [<HttpGet>]
    member _.GetPersonsByIdAsync([<FromRoute>]id) =
        task {
            let result = personsRepository.GetPersonById(id) 

            let httpResult = 
                match result with
                |Some(r) ->  this.Ok(r) :> IActionResult
                |_ ->  this.NotFound(id) :> IActionResult

            return httpResult
        }

    [<Route("persons")>]
    [<HttpPost>]
    member _.InsertPerson([<FromBody>] personDto) =
       task {
            let id = personsRepository.InsertPerson(personDto)
            return this.Ok({ personDto with Id = id })
        }

    
    [<Route("persons")>]
    [<HttpPut>]
    member _.UpdatePerson([<FromBody>] personDto) =
       task {
            personsRepository.UpdatePerson(personDto)
            return this.Ok(personDto)
        }

    
    [<Route("persons/{id}")>]
    [<HttpDelete>]
    member _.DeletePerson([<FromRoute>] id) =
       task {
            personsRepository.DeletePerson(id)
            return this.StatusCode(204)
        }

        

